### PR TITLE
Fix/selfhost mcp sidecar token

### PR DESCRIPTION
**Pull request by tylerwarburton**
Self-hosted MCP could list tools but sibling API calls failed the sidecar transport gate because most registry fetches omitted the local token. This repair applies the existing origin-scoped transport policy to every RPC, company-intelligence and NLP fetch.

`X-WorldMonitor-Local-Token` is attached only when the request targets the execution context's exact loopback origin. Token-bearing redirects are rejected so fetch cannot forward the custom header to another origin. Operator-key headers and Pro/user-key HMAC signing remain unchanged, including signed POST bodies, timeouts and response handling.

The contributor's commits are preserved, with current `main` merged. The Compose key wiring and self-hosting guide remain; the guide now reflects nginx's dedicated transport header. The old unconditional bearer injection is removed. This addresses the prior review's credential-forwarding concern.

Validation for `5d208158f`:

- 1,938 MCP tests passed, including six-tool × three-identity × local/remote coverage, exact HMAC verification, origin exclusions, a real HTTP redirect test and a real sidecar-gate test.
- 489 sidecar tests passed.
- API typecheck and Convex call audit passed.
- Markdown lint and an isolated shell check of the documented key setup passed. The example now defines the same key used by curl.
- Full lint and the normal pre-push gates passed. CI caught an undeclared name in the source-only quota fixture; the final commit declares its stub, and all 18 quota tests pass.
- Registry source guard confirms every sibling fetch uses the shared policy and receives execution context. The quota fan-out guard counts the policy helper; quota weights are unchanged.
- Sequential `ce-code-review` passes found no remaining actionable code findings.

The real sidecar test uses a fixture endpoint. Full Docker deployment and production acceptance are not claimed. Formal approval and merge remain separate from this code review.

Final exact-head CI: 34 GitHub check runs passed, 11 skipped; CodeRabbit and the aggregate `gate` status passed. Vercel remains failed because its fork preview requires deployment authorization. No deployment was authorized or attempted. GitHub reports `MERGEABLE / UNSTABLE`.

## Summary by CodeRabbit

- **New Features**
  - Self-hosted deployments can authenticate MCP and REST requests with configured operator keys.
  - MCP downstream requests support secure local authentication, identity propagation, and redirect protection.
  - Self-hosting documentation now includes API key configuration and usage.

- **Bug Fixes**
  - Improved local-host detection, including bracketed IPv6 loopback addresses.
  - Standardized authentication handling across MCP tools and downstream services.
  - Nested local API requests no longer consume upstream capacity unnecessarily.

- **Tests**
  - Added coverage for authentication, transport security, redirects, request propagation, concurrency, and tool request counting.

### CODE REVIEW

**Review by rajpratham1 — DISMISSED** [review]
This pull request provides a thoughtful fix for self-hosted MCP authentication by separating the sidecar transport token from user authentication, restoring X-WorldMonitor-Key authentication behind the Docker sidecar, documenting the deployment flow, and adding strong regression tests that validate both functionality and security boundaries. However, one blocking concern remains. The newly introduced transport Authorization header is added whenever LOCAL_API_TOKEN is configured, but its application is not restricted to loopback or internal sidecar destinations. Although the implementation comments state that the token is "only ever sent to our own loopback sidecar," the code does not enforce that invariant. Restricting the header injection to loopback addresses (or another explicit internal target) would prevent accidental forwarding of an internal credential if future tool implementations issue requests to external URLs. This should be addressed before approval.

**Review by koala73 — DISMISSED** [review]
## Review (not ready)

**CONFLICTING** with current `api/mcp/auth.ts`.

`resolveAuthContext` ignoring the nginx-injected `LOCAL_API_TOKEN` bearer so `X-WorldMonitor-Key` can work is the right self-host fix, and the tests for “token alone is still 401” are good.

**Do not attach `Authorization: Bearer ${LOCAL_API_TOKEN}` to every `buildAuthHeaders` result.** That helper is used for many tool fetches. There is no loopback/host check, so a self-hosted tool that ever targets a non-sidecar URL would leak the transport token. Restrict the extra header to the sidecar origin (or only to URLs that are already same-origin to the sidecar).

Rebase, add that restriction + a test, then this can be re-reviewed.

File: src-tauri/sidecar/local-api-server.mjs [hunk 1]
```diff
@@ -181,14 +187,16 @@ function makePinnedLookup(address, family = 4) {
 }
 
 function registerSidecarAllowedPrivateFetchOrigins(port, extraOrigins = []) {
-  const origins = [
+  const selfOrigins = [
```
[the remark this answers is not in the document]
Comment (koala73) [reply src-tauri/sidecar/local-api-server.mjs:193]: > If `options.port` or `LOCAL_API_PORT` is `80`, the sidecar registers `http://127.0.0.1:80`. `ipv4Fetch` compares this value with `url.origin`, which is `http://127.0.0.1`.

Addressed in 9723cd5d08: the self origins are now built through `new URL(...).origin`, so a default port compares equal to `url.origin` for both the private-fetch allowlist and the self-call semaphore exemption. No test binds port 80 (needs root in CI); the normalization is a URL-spec fact, and the existing #5441/#5449 semaphore tests still pass (sidecar suite 490/490).

File: api/mcp/auth.ts [hunk 2]
```diff
@@ -103,12 +103,25 @@ export async function buildAuthHeaders(
...
+    : {};
+
   if (context.kind === 'env_key' || context.kind === 'user_key') {
     // user_key (#4859): the downstream REST gateway validates the raw key
     // itself (Convex hash lookup + the #4611 apiAccess gate + per-account
     // limits), so usage attributes to the key owner exactly like a direct
     // REST call — no internal-HMAC identity smuggling needed.
-    return { 'X-WorldMonitor-Key': context.apiKey };
+    return { 'X-WorldMonitor-Key': context.apiKey, ...transportAuth };
   }
```
[the remark this answers is not in the document]
Comment (koala73) [reply api/mcp/auth.ts]: > When `LOCAL_API_TOKEN` is set in the container, `Authorization: Bearer ${sidecarToken}` is appended to every downstream tool fetch regardless of the target URL.

Not addressing here: this code no longer exists. `api/mcp/auth.ts` at the current head has no transport-token handling at all. The token is attached only in `api/mcp/downstream.ts` (`buildMcpDownstreamHeaders` / `fetchMcpDownstream`) behind three gates: the inbound host must be loopback, the target origin must equal the execution context's own downstream origin, and the target hostname must be loopback; a token-bearing fetch also refuses redirects. Pinned by the origin-exclusion matrix and the real two-server redirect test in `tests/mcp-registry-downstream.test.mjs`.
